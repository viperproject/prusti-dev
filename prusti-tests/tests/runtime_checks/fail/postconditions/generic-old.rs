//@run: 101
use prusti_contracts::*;

fn main() {
    let mut x = GenericStruct { x: 40};
    let mut y = GenericStruct { x: 32};
    let c = x.stuff(&mut y);
}

#[derive(Clone)]
struct GenericStruct<T> {
    pub x: T,
}

impl<T: PartialEq + Copy + std::ops::Add<Output=T>> GenericStruct<T> {
    #[trusted]
    // fails because of the +1
    #[insert_runtime_check]
    #[ensures(self.x == old(other.x))]
    #[insert_runtime_check]
    #[ensures(other.x == old(self.x))]
    pub fn stuff(&mut self, other: &mut Self) {
        std::mem::swap(&mut self.x, &mut other.x);
        // make check fail:
        self.x = self.x + other.x;
    }
}

