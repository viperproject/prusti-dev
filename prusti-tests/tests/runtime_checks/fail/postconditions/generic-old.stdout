check function prusti_post_check_item_stuff is performed
check function prusti_post_check_item_stuff is performed
